### DISCUSSION TITLE
Deployment options for self-hosted operation

### DISCUSSION THREAD

**Discussion by rhasselbusch** [Shelf-nu/shelf.nu#1231] [General]
Hi together,

i would love to evaluate Shelf for different needs (some prvate, some business-related).
Because my business has to do with governmental operations, it is strictly forbidden to save data in the cloud / on cloud services.

I am now wondering what priority the provision of a simpler deployment without the use of cloud services such as Supabase has for you, as I am not able to evaluate the product with the current status and the reddit posts i found tell me that i'm not the only one.

Best regards
Rolf

**Comment by DonKoko** [comment]
hello @rhasselbusch. Thanks for opening a discussion and your questions. Let me address the points you are raising.

>  Because my business has to do with governmental operations, it is strictly forbidden to save data in the cloud / on cloud services.

This is not the first case we hear that. You can of course self host it or if you have budget, get in touch with our sales and this is something we can offer to you as a service. We can setup shelf on your own servers, as long as they fit the requirements.

> I am now wondering what priority the provision of a simpler deployment without the use of cloud services such as Supabase has for you, as I am not able to evaluate the product with the current status and the reddit posts i found tell me that i'm not the only one.

Self-hosted is meant for users who have knowledge about setting it up themselves. Supabase is open-source and fully available for self-hosting, however giving guides on self hosting Supabase is outside the scope of shelf, that is why our documentation covers using the hosted version. If you have the expertise you can easily setup Supabase locally on the server that runs shelf. We have some community members that have done this. You can read about it in the Supabase documentation: https://supabase.com/docs/guides/self-hosting

I hope this make things a bit more clear. Please let us know if you have further questions.

**Comment by sochartetu** [comment]
Same question here, why do we need 2 databases ? And how to populate Supabase DB ?

**Reply by DonKoko** [reply]
Hey. I am not sure I understand your question. You don't need 2 databases. The guide simply shows how to setup with production and staging which is obviously not a requirement.

> And how to populate Supabase DB ?

Simply do `npm run setup` once you have all your stuff setup and it will run the migrations.

**Comment by sochartgit** [comment]
I am running Shelf under Docker with the prebuilt image  ghcr.io/shelf-nu/shelf.nu:latest
I have a stack with 2 services (shelf and a PSQL DB related to DIRECT_URL
DATABASE_URL refers to Supabase DB.

If I connect to shelf console and run:  npm run setup, it completes without error with the following message:
![image](https://github.com/user-attachments/assets/bf8f76b4-115b-4b5d-9e30-d5b2219f5398)

After that, Supabase DB public schema is still empty and I am not able to create a user account from the dashboard. It's accessed through a tunnel and has a proper SSL cert.
![image](https://github.com/user-attachments/assets/88204bb0-4a99-4b24-b2d5-892822cea115)


More than this, in the documentation we can read that:
![image](https://github.com/user-attachments/assets/744c5e21-d2ed-4401-8153-4821caf07554)

The npm run db:deploy-migration states it happened without any error but again, nothing is migrated to Supabase.
The Connect button on Supabase.com publishes a string like that:

postgresql://postgres.XXXXXXXXXXXXXXXX:[YOUR-PASSWORD]@aws-0-ca-central-1.pooler.supabase.com:6543/postgres

Whereas shelf-nu docs states the  {DB_NAME} variable has to be the same for DATABASE_URL and for DIRECT_URL.
And yes, I've tried with "postgres" and "shelf". Same thing = nothing is written in the Supabase DB.

I am lost, really. Thanks for any help you can provide.

**Reply by DonKoko** [reply]
Hey, let me try to help you. What confuses me is this part:
>  I have a stack with 2 services (shelf and a PSQL DB related to DIRECT_URL

Do I understand that you are running postgres locally? If that is the case then this is the problem. You should be connecting to the supabase postgres. 

> Whereas shelf-nu docs states the {DB_NAME} variable has to be the same for DATABASE_URL and for DIRECT_URL.
And yes, I've tried with "postgres" and "shelf". Same thing = nothing is written in the Supabase DB.

The only difference in `DATABASE_URL` and `DIRECT_URL` is the port. The rest is completely the same. DATABASE_URL is for connecting via supavisor pooler and DIRECT_URL is for prisma to run migrations. That is the only thing it does.

> The npm run db:deploy-migration states it happened without any error but again, nothing is migrated to Supabase.
The Connect button on Supabase.com publishes a string like that:
postgresql://postgres.XXXXXXXXXXXXXXXX:[YOUR-PASSWORD]@aws-0-ca-central-1.pooler.supabase.com:6543/postgres

That is correct. You can place this url in the DATABASE_URL and replace just the user and password. A few things:
1. You can find the user in `Project Settings > Database > Connection Parameters`. Make sure to get the whole user and notice it looks something like this: `postgres.yourprojectref`
2. Then you simply need to insert the password you have created.
3. For the `DIRECT_URL` it is exactly the same, just change the port from `6543` to `5432`. In the `Connection String` on supabase you can change the mode to see the session connection string.

Hope that helps. Feel free to also join our discord and ask there. I can also help you on discord in a bit faster manner if that is what you prefer.

**Comment by DominatorFTW** [comment]
Nice nice
